File: app01_java_nft/backend/src/main/java/com/gandarych/nftcerts/api/CertificateController.java
package com.gandarych.nftcerts.api;

import com.gandarych.nftcerts.certificate.Certificate;
import com.gandarych.nftcerts.certificate.CertificateService;
import com.gandarych.nftcerts.pdf.CertificatePdfService;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * GET /api/certificates/{tokenId} and GET /api/certificates/{tokenId}/pdf — read-only access to a
 * minted certificate's data and its downloadable PDF certificate of authenticity.
 */
@RestController
@RequestMapping("/api/certificates")
public class CertificateController {

    private final CertificateService certificateService;
    private final CertificateDtoMapper certificateDtoMapper;
    private final CertificatePdfService certificatePdfService;

    public CertificateController(CertificateService certificateService, CertificateDtoMapper certificateDtoMapper,
                                  CertificatePdfService certificatePdfService) {
        this.certificateService = certificateService;
        this.certificateDtoMapper = certificateDtoMapper;
        this.certificatePdfService = certificatePdfService;
    }

    @GetMapping("/{tokenId}")
    public ResponseEntity<CertificateDto> getCertificate(@PathVariable Long tokenId) {
        Certificate certificate = certificateService.getCertificate(tokenId);
        return ResponseEntity.ok(certificateDtoMapper.toDto(certificate));
    }

    @GetMapping("/{tokenId}/pdf")
    public ResponseEntity<byte[]> getCertificatePdf(@PathVariable Long tokenId) {
        Certificate certificate = certificateService.getCertificate(tokenId);
        byte[] pdf = certificatePdfService.generate(
                certificate,
                certificateDtoMapper.buildEtherscanUrl(certificate),
                certificateDtoMapper.buildOpenSeaUrl(certificate),
                certificateDtoMapper.buildRaribleUrl(certificate));

        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_PDF)
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"certificate-" + tokenId + ".pdf\"")
                .body(pdf);
    }
}

File: app02_cpp_nft/backend/src/api/CertificateController.cpp
#include "api/CertificateController.h"

#include "error/Exceptions.h"

#include <drogon/HttpAppFramework.h>

namespace nftcerts::api {

namespace {

uint64_t parseTokenId(const std::string& raw) {
    try {
        size_t consumed = 0;
        uint64_t value = std::stoull(raw, &consumed);
        if (consumed != raw.size()) {
            throw error::CertificateNotFoundException(-1);
        }
        return value;
    } catch (const std::invalid_argument&) {
        throw error::CertificateNotFoundException(-1);
    } catch (const std::out_of_range&) {
        throw error::CertificateNotFoundException(-1);
    }
}

}  // namespace

void registerCertificateRoutes(certificate::CertificateService& certificateService,
                                CertificateDtoMapper& certificateDtoMapper,
                                pdf::CertificatePdfService& certificatePdfService) {
    drogon::app().registerHandler(
        "/api/certificates/{1}",
        [&certificateService, &certificateDtoMapper](
            const drogon::HttpRequestPtr&, std::function<void(const drogon::HttpResponsePtr&)>&& callback,
            std::string tokenIdRaw) {
            uint64_t tokenId = parseTokenId(tokenIdRaw);
            certificate::Certificate cert = certificateService.getCertificate(tokenId);
            callback(drogon::HttpResponse::newHttpJsonResponse(certificateDtoMapper.toDto(cert)));
        },
        {drogon::Get});

    drogon::app().registerHandler(
        "/api/certificates/{1}/pdf",
        [&certificateService, &certificateDtoMapper, &certificatePdfService](
            const drogon::HttpRequestPtr&, std::function<void(const drogon::HttpResponsePtr&)>&& callback,
            std::string tokenIdRaw) {
            uint64_t tokenId = parseTokenId(tokenIdRaw);
            certificate::Certificate cert = certificateService.getCertificate(tokenId);

            std::vector<unsigned char> pdf = certificatePdfService.generate(
                cert, certificateDtoMapper.buildEtherscanUrl(cert), certificateDtoMapper.buildOpenSeaUrl(cert),
                certificateDtoMapper.buildRaribleUrl(cert));

            auto response = drogon::HttpResponse::newHttpResponse();
            response->setContentTypeCode(drogon::CT_APPLICATION_OCTET_STREAM);
            response->addHeader("Content-Type", "application/pdf");
            response->addHeader("Content-Disposition",
                                 "attachment; filename=\"certificate-" + std::to_string(tokenId) + ".pdf\"");
            response->setBody(std::string(pdf.begin(), pdf.end()));
            callback(response);
        },
        {drogon::Get});
}

}  // namespace nftcerts::api
